galley/pkg/mcp/server/server.go

```diff
@@ -0,0 +1,285 @@
+	peerInfo, ok := peer.FromContext(stream.Context())
+	peerAddr := "0.0.0.0"
+	if ok {
+		peerAddr = peerInfo.Addr.String()
+	}
+
+	con := &connection{
+		stream:    stream,
+		peerAddr:  peerAddr,
+		requestC:  make(chan *xdsapi.DiscoveryRequest),
+		responseC: make(chan *WatchResponse, 0),
+		watches:   make(map[string]*watch),
+		watcher:   s.watcher,
+		id:        atomic.AddInt64(&s.nextStreamID, 1),
+	}
+
+	var prettyTypes []string
+	for _, typ := range s.supportedTypes {
+		con.watches[typ] = &watch{}
+
+		// strip the common type.googleapis.com prefix
+		segs := strings.Split(typ, "/")
+		if len(segs) == 2 {
+			prettyTypes = append(prettyTypes, segs[1])
+		}
+	}
+
+	mcpLog.Infof("MCP: connection %v: NEW, supported types: %v", con, prettyTypes)
+	return con
+}
+
+// StreamAggregatedResources implements bidirectional streaming method for MCP (ADS).
+func (s *Server) StreamAggregatedResources(stream discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer) error { // nolint: lll
+	con := s.newConnection(stream)
+	defer con.close()
+	go con.receive()
+
+	for {
+		select {
+		case resp, more := <-con.responseC:
+			if !more {
+				return status.Errorf(codes.Unavailable, "watch failed")
+			}
+			if err := con.pushServerResponse(resp); err != nil {
+				return err
+			}
+		case req, more := <-con.requestC:
+			if !more {
+				return con.reqError
+			}
+			if err := con.processClientRequest(req); err != nil {
+				return err
+			}
+		case <-stream.Context().Done():
+			mcpLog.Debugf("MCP: connection %v: stream done, err=%v", con, stream.Context().Err())
+			return stream.Context().Err()
+		}
+	}
+}
+
+func (con *connection) String() string {
+	return fmt.Sprintf("{addr=%v id=%v}", con.peerAddr, con.id)
+}
+
+func (con *connection) send(resp *WatchResponse) (string, error) {
+	resources := make([]types.Any, 0, len(resp.Resources))
+	for _, resource := range resp.Resources {
+		data, err := types.MarshalAny(resource)
+		if err != nil {
+			mcpLog.Errorf("MCP: connection %v: internal error: %v", con, err)
+			return "", err
+		}
+		resources = append(resources, *data)
+	}
+	msg := &xdsapi.DiscoveryResponse{
+		VersionInfo: resp.Version,
+		Resources:   resources,
+		TypeUrl:     resp.TypeURL,
+	}
+
+	// increment nonce
+	con.streamNonce = con.streamNonce + 1
+	msg.Nonce = strconv.FormatInt(con.streamNonce, 10)
+	if err := con.stream.Send(msg); err != nil {
+		return "", err
+	}
+	mcpLog.Infof("MCP: connection %v: SEND version=%v nonce=%v", con, resp.Version, msg.Nonce)
+	return msg.Nonce, nil
+}
+
+func (con *connection) receive() {
+	defer close(con.requestC)
+	for {
+		req, err := con.stream.Recv()
+		if err != nil {
+			if status.Code(err) == codes.Canceled || err == io.EOF {
+				mcpLog.Infof("MCP: connection %v: TERMINATED %q", con, err)
+				return
+			}
+			mcpLog.Errorf("MCP: connection %v: TERMINATED with errors: %v", con, err)
+			con.reqError = err
```

Comment:
A comment here would be nice to indicate that this is set prior to closing of the channel, and listener (should) access the error after the channel closure.